File: kotlin/src/main/kotlin/aether/reputation/ReputationGossipService.kt
// SPDX-License-Identifier: MIT

package aether.reputation

import aether.protocol.PacketType
import aether.security.NodeReputationService
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Signed peer-to-peer reputation-score propagation service.
 *
 * Implements the gossip half of the Aether security hardening layer:
 * - **Broadcast**: serialize, sign, and fan-out a [ReputationUpdatePayload] to
 *   all directly-connected peers.
 * - **Receive**: verify signature, apply freshness / echo guards, weight the
 *   incoming delta by the reporter's own reputation, and commit the effective
 *   delta to [NodeReputationService].
 *
 * Wire format: `PacketType.ReputationUpdate` (value 52), payload is
 * UTF-8 JSON (snake_case keys via [SerialName] annotations).
 *
 * Freshness window: ±5 minutes. Stale or future-dated packets are silently
 * dropped. Self-echo guard: the node never applies its own re-broadcast.
 *
 * Effective delta formula:
 * ```
 * effectiveDelta = clamp(scoreDelta, −1, 1) × reporterReputation
 * ```
 * Unknown reporters default to reputation 1.0 (benefit of the doubt).
 */
class ReputationGossipService(
    private val meshSender: MeshSender,
    private val packetSigner: PacketSigner,
    private val reputation: NodeReputationService,
) {

    // ── Injectable abstractions ───────────────────────────────────────────────

    interface MeshSender {
        val localUhid: String
        fun broadcast(packet: GossipPacket): Int
    }

    interface PacketSigner {
        fun sign(packet: GossipPacket)
        fun verify(packet: GossipPacket, senderPublicKey: ByteArray): Boolean
    }

    // ── Wire types ────────────────────────────────────────────────────────────

    @Serializable
    data class ReputationUpdatePayload(
        @SerialName("reporter_uhid") val reporterUhid: String,
        @SerialName("target_uhid")   val targetUhid: String,
        @SerialName("score_delta")   val scoreDelta: Double,
        @SerialName("timestamp_ms")  val timestampMs: Long,
        val reason: String,
    )

    data class GossipPacket(
        val packetType: Byte,
        val sourceUhid: String,
        val destinationUhid: String,
        val ttl: Int,
        val payload: ByteArray,
        val timestampMs: Long,
        var signature: ByteArray = ByteArray(0),
        var packetNonce: ByteArray = ByteArray(0),
    )

    // ── Public API ────────────────────────────────────────────────────────────

    /**
     * Build, sign, and broadcast a [ReputationUpdatePayload] to all
     * directly-connected peers.
     *
     * [scoreDelta] is clamped to [−1, 1] before serialisation.
     * Returns the broadcast fan-out count.
     */
    fun broadcastReputationUpdate(targetUhid: String, scoreDelta: Double, reason: String): Int {
        val clamped = scoreDelta.coerceIn(-1.0, 1.0)
        val nowMs = System.currentTimeMillis()
        val payload = ReputationUpdatePayload(
            reporterUhid = meshSender.localUhid,
            targetUhid   = targetUhid,
            scoreDelta   = clamped,
            timestampMs  = nowMs,
            reason       = reason,
        )
        val payloadBytes = Json.encodeToString(payload).toByteArray(Charsets.UTF_8)
        val packet = GossipPacket(
            packetType       = PacketType.ReputationUpdate.value,
            sourceUhid       = meshSender.localUhid,
            destinationUhid  = "*",
            ttl              = 3,
            payload          = payloadBytes,
            timestampMs      = nowMs,
        )
        packetSigner.sign(packet)
        return meshSender.broadcast(packet)
    }

    /**
     * Process an inbound gossip packet.
     *
     * Returns `true` if the packet was accepted and the weighted delta was
     * applied to the local reputation service. Returns `false` (no-op) if any
     * guard check fails.
     *
     * Guards (in order):
     * 1. Packet type must be [PacketType.ReputationUpdate].
     * 2. Signature must be valid.
     * 3. Payload must be valid JSON.
     * 4. Freshness: |now − timestampMs| ≤ 5 minutes.
     * 5. Non-empty reporter and target UHIDs.
     * 6. Own-echo: reporter must not be this node.
     * 7. Apply `effectiveDelta = clamp(scoreDelta, −1, 1) × reporterReputation`.
     */
    fun handleGossipPacket(packet: GossipPacket, senderPublicKey: ByteArray): Boolean {
        // 1. Type guard
        if (packet.packetType != PacketType.ReputationUpdate.value) return false
        // 2. Signature
        if (!packetSigner.verify(packet, senderPublicKey)) return false
        // 3. Parse payload
        val payload = try {
            Json.decodeFromString<ReputationUpdatePayload>(
                packet.payload.toString(Charsets.UTF_8)
            )
        } catch (_: Exception) {
            return false
        }
        // 4. Freshness (±5 min)
        val nowMs = System.currentTimeMillis()
        if (kotlin.math.abs(nowMs - payload.timestampMs) > FRESHNESS_WINDOW_MS) return false
        // 5. Non-empty UHIDs
        if (payload.reporterUhid.isEmpty() || payload.targetUhid.isEmpty()) return false
        // 6. Own-echo guard
        if (payload.reporterUhid == meshSender.localUhid) return false
        // 7. Weighted delta
        val reporterScore = reputation.getReputationScore(payload.reporterUhid)
        val clamped = payload.scoreDelta.coerceIn(-1.0, 1.0)
        val effective = clamped * reporterScore
        reputation.applyWeightedDelta(payload.targetUhid, effective)
        return true
    }

    private companion object {
        const val FRESHNESS_WINDOW_MS = 5L * 60L * 1000L // 5 minutes
    }
}

File: swift/Sources/AetherMeshProtocol/Reputation/ReputationGossipService.swift
// SPDX-License-Identifier: MIT

import Foundation

// MARK: - Payload

/// Wire-format payload for a reputation-gossip broadcast.
///
/// JSON keys use snake_case to match the cross-language Aether wire convention.
public struct ReputationUpdatePayload: Codable {
    public let reporterUhid: String
    public let targetUhid: String
    public let scoreDelta: Double
    public let timestampMs: Int64
    public let reason: String

    enum CodingKeys: String, CodingKey {
        case reporterUhid = "reporter_uhid"
        case targetUhid   = "target_uhid"
        case scoreDelta   = "score_delta"
        case timestampMs  = "timestamp_ms"
        case reason
    }

    public init(
        reporterUhid: String,
        targetUhid: String,
        scoreDelta: Double,
        timestampMs: Int64,
        reason: String
    ) {
        self.reporterUhid = reporterUhid
        self.targetUhid   = targetUhid
        self.scoreDelta   = scoreDelta
        self.timestampMs  = timestampMs
        self.reason       = reason
    }
}

// MARK: - Protocols

/// Minimal broadcast abstraction consumed by ``ReputationGossipService``.
///
/// Intentionally distinct from the routing-layer ``MeshSender`` so that the
/// gossip service can be injected with a thin adapter over any transport.
public protocol GossipMeshSender: Sendable {
    /// The local node's UHID, used as `reporterUhid` on outbound gossip.
    var localUhid: String { get }

    /// Broadcast `packet` to all reachable peers; returns the fan-out count.
    func broadcast(packet: MeshPacket) async throws -> Int
}

/// Signs and verifies ``MeshPacket`` instances for gossip traffic.
public protocol GossipPacketSigner: Sendable {
    /// Populate `packet.packetNonce`, `packet.timestampMs`, and
    /// `packet.signature` in place.
    func sign(packet: inout MeshPacket) async throws

    /// Return `true` iff the packet signature is valid for `senderPublicKey`.
    func verify(packet: MeshPacket, senderPublicKey: [UInt8]) async throws -> Bool
}

// MARK: - Service

/// Gossip layer for distributing reputation observations across the mesh.
///
/// - Broadcasts a signed ``ReputationUpdatePayload`` to all peers (TTL = 3).
/// - Accepts incoming gossip, verifies authenticity and freshness, then applies
///   a *weighted* score delta — scaled by the reporter's own local reputation
///   score — to prevent low-trust nodes from injecting large adjustments.
///
/// Thread-safety: implemented as a Swift `actor`; all state mutations run on
/// the actor's executor.
public actor ReputationGossipService {

    // MARK: - Dependencies

    private let sender:     any GossipMeshSender
    private let signing:    any GossipPacketSigner
    private let reputation: NodeReputationService

    // MARK: - Constants

    /// Maximum allowed clock skew between gossip payload timestamp and local
    /// wall-clock time before a packet is considered stale.
    private static let freshnessWindowMs: Int64 = 5 * 60 * 1_000   // 5 minutes

    /// TTL applied to every outbound reputation-gossip broadcast.
    private static let gossipTtl: Int32 = 3

    // MARK: - Init

    public init(
        sender:     any GossipMeshSender,
        signing:    any GossipPacketSigner,
        reputation: NodeReputationService
    ) {
        self.sender     = sender
        self.signing    = signing
        self.reputation = reputation
    }

    // MARK: - Outbound

    /// Broadcast a reputation observation about `targetUhid` to all peers.
    ///
    /// `scoreDelta` is clamped to [−1, 1] before encoding so that no single
    /// broadcast can carry an extreme adjustment.
    @discardableResult
    public func broadcastReputationUpdate(
        targetUhid: String,
        scoreDelta: Double,
        reason: String
    ) async throws -> Int {
        let clamped = max(-1.0, min(1.0, scoreDelta))

        let nowMs = Int64(Date().timeIntervalSince1970 * 1_000)
        let payload = ReputationUpdatePayload(
            reporterUhid: sender.localUhid,
            targetUhid:   targetUhid,
            scoreDelta:   clamped,
            timestampMs:  nowMs,
            reason:       reason
        )

        let payloadData = try JSONEncoder().encode(payload)

        var packet = MeshPacket(
            type:            .reputationUpdate,
            sourceUhid:      sender.localUhid,
            destinationUhid: "*",
            ttl:             Self.gossipTtl,
            payload:         payloadData
        )

        try await signing.sign(packet: &packet)
        return try await sender.broadcast(packet: packet)
    }

    // MARK: - Inbound

    /// Process an inbound gossip packet received from a peer.
    ///
    /// Returns `true` if the packet was accepted and the reputation store was
    /// updated; `false` for any rejection (wrong type, bad signature, stale,
    /// malformed, own-echo, etc.).
    public func handleGossipPacket(
        packet: MeshPacket,
        senderPublicKey: [UInt8]
    ) async throws -> Bool {

        // 1. Correct packet type guard.
        guard packet.type == .reputationUpdate else { return false }

        // 2. Signature verification.
        let valid = try await signing.verify(packet: packet,
                                             senderPublicKey: senderPublicKey)
        guard valid else { return false }

        // 3. Decode payload.
        guard let payload = try? JSONDecoder().decode(
            ReputationUpdatePayload.self, from: packet.payload
        ) else { return false }

        // 4. Freshness check — reject if timestamp drifts more than 5 minutes.
        let nowMs = Int64(Date().timeIntervalSince1970 * 1_000)
        guard abs(nowMs - payload.timestampMs) <= Self.freshnessWindowMs else {
            return false
        }

        // 5. Non-empty reporter and target UHIDs.
        guard !payload.reporterUhid.isEmpty, !payload.targetUhid.isEmpty else {
            return false
        }

        // 6. Own-echo guard — never apply gossip we ourselves originated.
        guard payload.reporterUhid != sender.localUhid else { return false }

        // 7. Weight the delta by the reporter's own reputation score.
        let reporterScore = await reputation.reputationScore(for: payload.reporterUhid)

        // 8. Clamp the raw delta from the payload.
        let clamped = max(-1.0, min(1.0, payload.scoreDelta))

        // 9. Scale by reporter trust.
        let effectiveDelta = clamped * reporterScore

        // 10. Apply to the target's score.
        await reputation.applyWeightedDelta(uhid: payload.targetUhid,
                                            weightedDelta: effectiveDelta)

        return true
    }
}
